### PR TITLE
Basic blog layout changes

### PR DESCRIPTION
**Pull request by kapowaz**
This PR contains a number of basic high-level layout changes:

* Remove all existing bootstrap CSS etc.
* Remove all presentational (in fact, _all_) classnames from existing markup
* Strip down existing markup templates to the essentials, so we can build up from there
* Add some basic SCSS mixins that will help with creating responsive layouts etc.

A future PR will finalise branding / colours etc. but for now this should be a decent starting point for principle layout concerns.

### CODE REVIEW

**Comment by mezis** [comment]
Why would we remove Bootstrap? Or rather, what would you replace it with?

Rationale: let's not reinvent a grid and basic styling again, pretty please?

**Comment by kapowaz** [comment]
Bootstrap provides masses and masses of stuff that the engineering blog simply doesn’t need. Worse, it imposes conventions of using presentational classnames. I’ve taken the decision that when building out a new style (!) for the blog, I would rather not fight against existing classnames (or try to shoehorn them into that style) but rather start with a relatively clean slate.

I can understand the concerns with not wanting to reinvent the wheel, but Bootstrap isn’t a wheel: it’s an SUV. We don’t need an SUV here, though — the content on the blog will be relatively simple stuff (I don’t think a grid layout is necessarily even going to be needed here).

I tend to take a rather different approach to writing CSS to a lot of the prevailing methods out there (OOCSS, BEM etc.) and I’m eager to see them more widely adopted because I’m convinced they’re better than those other methods, and work well. Assuming of course there is no serious objection, I thought that the engineering blog would be a good place to demonstrate these techniques, let people see how they work and encourage a conversation about how we write CSS in general going forwards. I’m happy to drive this conversation (in fact I’d relish the chance to!) but I’m aware that other people will probably have strong views about this too!

**Comment by mezis** [comment]
> I tend to take a rather different approach to writing CSS to a lot of the prevailing methods out there (OOCSS, BEM etc.)

I'm actually a big fan of (extremely strict) OOCSS or BEM.

> Assuming of course there is no serious objection, I thought that the engineering blog would be a good place to demonstrate these techniques

Indeed. I'll leave you to it then; worst case, as you hint to, a full restyle of this app wouldn't be a nightmare.

**Comment by kapowaz** [comment]
Looks like I have my work cut out for me to do the persuadin’, then! It’s a few years old now, but the best summary of my feelings about CSS authoring are summed up in [something I wrote here](http://www.kapowaz.net/articles/cargo-cult-css). Feel free to read and let me know what you think 😁

**Comment by kapowaz** [comment]
I’ve rebased this branch on top of your article and made some changes to it, in order to test layouts with the new CSS. The [changes to your article](https://github.com/deliveroo/deliveroo.engineering/pull/28/commits/a3db9850533838b38dce553bb14395c42e59fa16) are essentially:

* Use `<aside>` markup around the aside blocks rather than blockquote indentation
* Use `<figure>` elements around images
* Remove all classnames! 😬 

The `<blockquote>` element felt like the wrong choice for these blocks, and using figure gives us a semantic wrapper that can be useful for other things (like including a `<caption>`).

**Comment by kapowaz** [comment]
img

**Comment by kapowaz** [comment]
Regarding [this comment](#issuecomment-242424245) I have since rebased this out so it’s no longer part of this PR. Nonetheless there are some changes that will need to be made to all existing and future markdown files to ensure they adhere to the new layout:

* Change blockquote (`> …`) indented blocks to be wrapped with an `<aside>` element
* Wrap images with a `<figure>` element
* Remove all inline class identifiers (e.g. `{: .dg-figure}`)

The idea is that your articles and guideline documents should be as **appearance-agnostic as possible**; articles shouldn’t need to know anything about how they’re being presented, just that they have the correct semantic markup for whatever subject material they’re dealing with.

File: _sass/local/_color.scss [hunk 1]
```diff
@@ -0,0 +1,25 @@
+@function color($key, $lighten: null, $darken: null, $transparency: null) {
+  @if map-has-key($colors, $key) {
+    // if a lighten argument was passed, mix that colour with white
+    @if $lighten != null {
```
Comment (sdorunga1) [inline _sass/local/_color.scss]: Can't you default to some usable values instead that are the identity values for these functions? I.e. if you make lighten 0 will it not just do nothing?

Comment (kapowaz) [inline _sass/local/_color.scss]: @sdorunga1 if the function only supported the presence/absence of a single argument that would work, but unfortunately we’d need to know that the value was null to ensure flow control drops down to the next condition (`@if $darken != null`). Unless I’m misunderstanding and there’s some better way of doing that (unfortunately SCSS doesn’t have any concept of a `switch` statement)?

File: _includes/footer.html [hunk 2]
```diff
@@ -1,12 +1,8 @@
...
-        <a href="https://creativecommons.org/licenses/by-sa/4.0/">
-          Creative Commons attribution-sharealike</a>.
-      </div>
-    </div>
-  </div>
+<footer>
+  <ul class="links">
+    <li><a href="{{site.data.routes.guidelines}}">Engineering Guidelines</a></li>
+    <li><a href="{{site.data.routes.design_blog}}" target="_blank">deliveroo.design</a></li>
+    <li><a href="{{site.data.routes.food_blog}}" target="_blank">Deliveroo food blog</a></li>
```
Comment (jphastings) [inline _includes/footer.html]: We have more than one food blog, I've raised #33 to address that we probably ought to be linking to them all somehoe.

Comment (kapowaz) [inline _includes/footer.html]: Yeah, I wasn’t sure exactly how this might look so for now the footer content is pretty much placeholder.

File: _layouts/post.html [hunk 3]
```diff
@@ -1,15 +1,18 @@
...
-
-  <header class="post-header">
-    <h1 class="dg-title" itemprop="name headline">{{ page.title }}</h1>
-    <p class="dg-subtitle"><time datetime="{{ page.date | date_to_xmlschema }}" itemprop="datePublished">{{ page.date | date: "%b %-d, %Y" }}</time>{% if page.author %} • <span itemprop="author" itemscope itemtype="http://schema.org/Person"><span itemprop="name">{{ page.author }}</span></span>{% endif %}</p>
+<article>
+  <header>
+    <h1>{{ page.title }}</h1>
+    <time datetime="{{ page.date | date_to_xmlschema }}" class="published">{{ page.date | date: "%A, %B %-d, %Y" }}</time>
+    {% if page.author %}
+      <p class="author">{{ page.author }}</p>
```
Comment (jphastings) [inline _layouts/post.html]: I have a branch which adds more information for authors here, I'll rebase on top of this once it's merged so we can augment this section :)

File: _sass/local/_color.scss [hunk 4]
```diff
@@ -0,0 +1,25 @@
+@function color($key, $lighten: null, $darken: null, $transparency: null) {
```
Comment (jphastings) [inline _sass/local/_color.scss]: Can we have an example usage, because a little bit of documentation is nice :)

Comment (kapowaz) [inline _sass/local/_color.scss]: Good idea!

**Comment by jphastings** [comment]
LGTM! I think you're good to merge :)

File: _sass/local/_darkly.scss [hunk 5]
```diff
@@ -0,0 +1,34 @@
+/*
```
Comment (lededje) [inline _sass/local/_darkly.scss]: What dis

File: _sass/local/content.scss [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+p.author {
+  color: color(light_grey);
+  margin-top: 0;
+} // p.author
+
+section + h2 {
+  margin-top: 50px;
+} // section + h2
+
+section#recent-posts {
```
Comment (lededje) [inline _sass/local/content.scss]: This feels like overkill. We typically avoid specifying the element, when using classes or ids. Is there a reason you specify them here?

File: _sass/local/content.scss [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+  } // article code.highlighter-rouge
+  
+  pre.highlight {
+    overflow: auto;
+  } // article pre.highlight
+  
+  sup {
+    a.footnote {
+      padding: 0 2px;
+    } // article sup a.footnote
```
Comment (lededje) [inline _sass/local/content.scss]: I get that these endings might be useful for large definition sets but on ones that are not but a few lines long it feels overkill.

File: _sass/local/content.scss (hunk 6 above)
Comment (kapowaz) [inline _sass/local/content.scss]: As I discussed with @mezis in the general comments on this PR, I’m trying something different with the CSS for the engineering blog. It’s a little beyond the scope of an inline comment here to discuss it in detail, but the TL;DR is that this is _intentionally_ specific, both for reasons of narrowly-defining the selector’s specificity, but also for self-documentation reasons.

File: _sass/local/_darkly.scss (hunk 5 above)
Comment (kapowaz) [inline _sass/local/_darkly.scss]: Darkly is a third-party colour scheme for the rouge syntax highlighter. The blog already had one in place, but its colour scheme wasn’t very nice. I’ve chosen the first half-decent dark-background syntax highlighting scheme I found, but I’m open to suggestions to use another one (or to tweak this, if preferred).

File: _sass/local/_color.scss (hunk 1 above)
Comment (lededje) [inline _sass/local/_color.scss]: Is this the same implementation being used on consumer web?

Comment (kapowaz) [inline _sass/local/_color.scss]: I don’t think it’s exactly the same implementation as that one wasn’t finalised when I started this. I’m happy to go back and update the implementation once this initial PR is sorted out, though.

File: _includes/header.html [hunk 8]
```diff
@@ -1,21 +1,14 @@
...
+        {% if _page.title and _page.nav %}
+          <li>
+            <a href="{{ _page.url | prepend: site.baseurl }}">{{ _page.title }}</a>
+          </li>
+        {% endif %}
+      {% endfor %}
+    </ul>
   </nav>
-</header>
+</header> <!-- /header#navigation -->
```
Comment (lededje) [inline _includes/header.html]: Doesn't `</header>` convey `<!-- /header#navigation -->` quite well on it's own?

File: _sass/local/_layout.scss [hunk 9]
```diff
@@ -0,0 +1,93 @@
+@mixin background-image-file($url, $background-size: null, $background-position: null, $background-repeat: no-repeat, $type: 'png', $retina: true) {
```
Comment (mezis) [inline _sass/local/_layout.scss]: Given how you're building a lot from scratch and want to demonstrate a better approach than Bootstrap: this break one of the Sandy Rules. Too many function parameters up there :)

(and this function is much too long)

I feel this at least warrants an explicit documentation comment?

File: _sass/local/content.scss [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+  margin-top: 50px;
+} // section + h2
+
+section#recent-posts {
+  @include auto-width(desktop_medium);
+
+  div.post {
+    h3 {
+      margin-bottom: 0.5em;
+    } // section#recent-posts div.post h3
```
Comment (mezis) [inline _sass/local/content.scss]: Lots of these "trailing" end-of-block comments here. They don't really help readability, and actively hurt if your editor does this for you. Could we avoid those?

File: _sass/local/default.scss [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+body {
+  display: flex;
+  flex-direction: column;
+  min-height: 100vh;
+  font-size: 62.5%;
+  color: color(dark_grey);
+  background-color: color(light_grey, 5%);
+  
+  main {
```
Comment (mezis) [inline _sass/local/default.scss]: Nesting `main` under `body` is unecessary.

File: _includes/header.html [hunk 12]
```diff
@@ -1,21 +1,14 @@
...
-              <a class="nav-link" href="{{ _page.url | prepend: site.baseurl }}">
-                {{ _page.title }}
-              </a>
-            </li>
-          {% endif %}
-        {% endfor %}
-      </ul>
-    </div>
+<header id="navigation">
+  <a href="{{ site.baseurl }}/">{{ site.title }}</a>
```
Comment (mezis) [inline _includes/header.html]: This link is semantically part of navigation, as it brings back to the home page.

Conveniently, if you nest it that way, you don't need the ID on the `header` element anymore. `header > nav` works just fine.
(assuming you insist on purely semantic, classless styling)

File: _includes/header.html (hunk 8 above)
Comment (kapowaz) [inline _includes/header.html]: Only if there’s only ever one `<header>` element on the page (which there won’t be in the case of post pages) — I’ve found it quite a useful debugging aid to be able to identify which element a closing tag corresponds to, which – thanks to the vagaries of templating systems’ indentation – is often quite tricky.

File: _sass/local/content.scss (hunk 10 above)
Comment (kapowaz) [inline _sass/local/content.scss]: The idea behind these is twofold:

1. To aid in determining which selector block you’re looking at, which is increasingly useful as these blocks grow in size
2. To reinforce the full generated selector size, which can aid with refactoring if the nesting/specificity is too big

I personally disagree that they harm readability — they’re just CSS comments, so should be no more visible/distracting than any other comment, plus they sit outside the closing brace.

 > actively hurt if your editor does this for you

Does your editor do this for you? Mine doesn’t 😢

File: _sass/local/default.scss (hunk 11 above)
Comment (kapowaz) [inline _sass/local/default.scss]: Technically it’s not necessary, but since there’s a flex box relationship between `body` and its children (`header`, `main`, and `footer`) I thought it was a useful way of illustrating that relationship.

File: _includes/header.html (hunk 12 above)
Comment (kapowaz) [inline _includes/header.html]: Yeah, I’ve actually changed the markup/layout of this in the next PR I’m working on to make it a sibling of the other `li` elements in the navigation — but that also involves adding the logo to the navigation so I’m keeping that out of this PR.

File: _sass/local/_layout.scss (hunk 9 above)
Comment (kapowaz) [inline _sass/local/_layout.scss]: Documentation would definitely be helpful here. I wasn’t sure how this could be broken up to have fewer parameters without ending up with something unwieldy, but then I remembered SASS’s `@content` [directive](http://sass-lang.com/documentation/file.SASS_REFERENCE.html#mixin-content) — maybe that could be made to work here, allowing some of the arguments to be just passed as a block?

File: _sass/local/_layout.scss [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+
+    @if $background-size {
+      background-size: $background-size;
+    }
+
+    background-repeat: $background-repeat;
+  }
+} // @mixin background-image-file
+
+@mixin image-replacement($url, $background-width, $background-height, $width: $background-width, $height: $background-height, $background-position: 0 0, $type: 'png') {
```
Comment (kapowaz) [inline _sass/local/_layout.scss]: I’m trying to see how this could be made to require fewer parameters, but I’m encountering problems with duplication. In trying to reduce the arguments (e.g. `$width` and `$height`) and having these passed in via the `@content` block, you end up duplicating these values in the callside, e.g:

```
@include image-replacement('/images/logo_engineering', 175px, 40px, $type: 'svg') {
  width: 175px;
  height: 40px;
  background-position: center;
};
```

I can more easily reduce the number of arguments here to 6, but that’s still ‘too many’. I’ll prepare a PR with the improvements (and documentation) and see how that sits?

File: _sass/local/content.scss (hunk 10 above)
Comment (mezis) [inline _sass/local/content.scss]: As much as I'm happy with experimenting with an alternate approach to CSS, I really do disagree with this.

1. These comments are almost guaranteed to fall out of sync.
2. "as these blocks grow in size" is a counter-argument: if the block doesn't fit half a screen, it needs a good refactor (it violates another of the Sandy Rules)
3. "reinforce the full generated selector size" is already mostly visible at the beginning of the declaration, though the nesting level and length of the subselector string.

File: _includes/header.html (hunk 8 above)
Comment (mezis) [inline _includes/header.html]: Agree with @lededje, and mirroring what I said on CSS comments: your editor should really highlight the matching element-start. I don't think there's a need for this, and it does indeed add some noise.

Comment (kapowaz) [inline _includes/header.html]: This is largely for debugging the _rendered_ markup, rather than what you see in the editor (if it was template code, say ERB, I’d use something like `<%# /header#navigation %>` instead anyway). I’m happy enough to remove it though!

File: _sass/local/content.scss (hunk 10 above)
Comment (kapowaz) [inline _sass/local/content.scss]: > These comments are almost guaranteed to fall out of sync.

It can happen, but the impact is small since they’re only comments. I would say the benefits of them being there _at all_ (even if sometimes they’re out of sync) outweigh them not being there.

> "as these blocks grow in size" is a counter-argument: if the block doesn't fit half a screen, it needs a good refactor (it violates another of the Sandy Rules)
>
> "reinforce the full generated selector size" is already mostly visible at the beginning of the declaration, though the nesting level and length of the subselector string.

I’ve found that writing out the generated selector often helps with offering up a path to refactoring, whereas a series of closing braces gives you very little help with that. Similarly the depth of nesting (and whether it fits on-screen at once) is often an indication that something might be better if refactored… but sometimes it can’t easily be (after all SCSS is _not_ Ruby, so a lot of the best practices from that domain don’t necessarily translate to this domain).

As a middle ground, perhaps adding a comment on the outermost nests would be acceptable?

Comment (martyphee) [inline _sass/local/content.scss]: I'd agree with Julien.  We shouldn't have heavily nested CSS in there.  One or two level at most.

Comment (kapowaz) [inline _sass/local/content.scss]: Ignoring the fact that orderweb is filled with examples of nesting many levels more deep than this, I’d still have to ask the question: _why not?_

The idea with this kind of nesting is that the generated CSS selectors reflect the underlying DOM structure they’re targeting. This has two benefits: they’re highly specific, and they’re self-documenting.

It may well be possible to refactor the nesting to be less specific, but as the intention with systems like BEM is to prevent unwanted side-effects, having more specificity in a modular chunk of (S)CSS written like this is actually a good thing. The further these selectors are nested, the more certain you can be that the same selector chain won’t match anything elsewhere unintentionally.

File: _includes/header.html (hunk 12 above)
Comment (kapowaz) [inline _includes/header.html]: I’ve now brought these changes in from the separate PR (minus the logo) to reduce the size of the diff between branches.

File: _layouts/post.html (hunk 3 above)
Comment (kapowaz) [inline _layouts/post.html]: I’m fleshing out all the currently-missing top level navigation pages with placeholder pages (e.g. `/articles/index.html`) so that they’re not 404s. Did you already create something similar for authors in your branch? I’m wondering if maybe these could all move into their own PR since it’s not really layout-related (and it’ll make it easier to get it merged sooner, too).
